### PR TITLE
Onboarding/task7

### PR DESCRIPTION
**Pull request by thalyscarrara**
## Description

implement repository pattern on google codelab

## Related Issues

[task07-onboarding](https://github.com/profusion/quickstart-android-jetpack-graphql/issues/207)

## Visual reference

https://user-images.githubusercontent.com/53837410/192619398-08f2ab36-2316-4d5a-a79f-54cd276511ee.mp4

### CODE REVIEW

**Review by mari1912 — APPROVED** [review]

File: RepositoryPattern-Starter/app/src/main/java/com/example/android/devbyteviewer/viewmodels/DevByteViewModel.kt [hunk 1]
```diff
@@ -45,24 +39,26 @@ class DevByteViewModel(application: Application) : AndroidViewModel(application)
...
+//    private val _playlist = MutableLiveData<List<DevByteVideo>>()
 
     /**
      * A playlist of videos that can be shown on the screen. Views should use this to get access
      * to the data.
      */
-    val playlist: LiveData<List<DevByteVideo>>
-        get() = _playlist
+//    val playlist: LiveData<List<DevByteVideo>>
+//        get() = _playlist
```
Comment (joaobertacchi) [inline RepositoryPattern-Starter/app/src/main/java/com/example/android/devbyteviewer/viewmodels/DevByteViewModel.kt:61]: Please don't commit commented out code. That's bad practice

File: RepositoryPattern-Starter/app/src/main/java/com/example/android/devbyteviewer/viewmodels/DevByteViewModel.kt [hunk 2]
```diff
@@ -95,25 +91,26 @@ class DevByteViewModel(application: Application) : AndroidViewModel(application)
...
         // TODO: Replace with a call to the refreshDataFromRepository9) method
-        refreshDataFromNetwork()
+        refreshDataFromRepository()
     }
 
     /**
      * Refresh data from the repository. Use a coroutine launch to run in a
      * background thread.
      */
     // TODO: Replace with the refreshDataFromRepository() method
```
Comment (joaobertacchi) [inline RepositoryPattern-Starter/app/src/main/java/com/example/android/devbyteviewer/viewmodels/DevByteViewModel.kt:101]: Please remove all the TODO's you have addressed.

File: RepositoryPattern-Starter/app/build.gradle [hunk 3]
```diff
@@ -93,8 +93,10 @@ dependencies {
     // arch components
     // ViewModel and LiveData
     def lifecycle_version = "2.4.1"
+    def lifecycle_livedata_version = "2.2.0"
 //    implementation "androidx.lifecycle:lifecycle-extensions:2.2.0"
```
Comment (joaobertacchi) [inline RepositoryPattern-Starter/app/build.gradle:97]: ditto

File: RepositoryPattern-Starter/app/src/main/java/com/example/android/devbyteviewer/viewmodels/DevByteViewModel.kt [hunk 4]
```diff
@@ -95,25 +91,26 @@ class DevByteViewModel(application: Application) : AndroidViewModel(application)
      */
     init {
         // TODO: Replace with a call to the refreshDataFromRepository9) method
```
Comment (joaobertacchi) [inline RepositoryPattern-Starter/app/src/main/java/com/example/android/devbyteviewer/viewmodels/DevByteViewModel.kt:93]: ditto

File: RepositoryPattern-Starter/app/src/main/java/com/example/android/devbyteviewer/viewmodels/DevByteViewModel.kt [hunk 5]
```diff
@@ -95,25 +91,26 @@ class DevByteViewModel(application: Application) : AndroidViewModel(application)
...
+            try {
+                videosRepository.refreshVideos()
+                _eventNetworkError.value = false
+                _isNetworkErrorShown.value = false
+
+            } catch (networkError: IOException) {
+                // Show a Toast error message and hide the progress bar.
+                if(playlist.value.isNullOrEmpty())
+                    _eventNetworkError.value = true
+            }
```
Comment (joaobertacchi) [inline RepositoryPattern-Starter/app/src/main/java/com/example/android/devbyteviewer/viewmodels/DevByteViewModel.kt:113]: You replaced
```
private fun funcName() = viewModelScope.launch
...
```
by

```
private fun funcName() {
   viewModelScope.launch {
...
```

What that intentional? Could you please explain that?

Comment (thalyscarrara) [reply RepositoryPattern-Starter/app/src/main/java/com/example/android/devbyteviewer/viewmodels/DevByteViewModel.kt:113]: I just followed the tutorial, what I understood from this part was, the idea of the repository is to create an abstraction layer to obtain the data, that is, whoever is going to use this data doesn't need to know where it came from

Comment (thalyscarrara) [reply RepositoryPattern-Starter/app/src/main/java/com/example/android/devbyteviewer/viewmodels/DevByteViewModel.kt:113]: now talking about the difference in the return of functions. In the first case, the declaration of a function that returns the reference of a coroutine happens and in the second case it executes the coroutine as soon as it is called

Comment (thalyscarrara) [reply RepositoryPattern-Starter/app/src/main/java/com/example/android/devbyteviewer/viewmodels/DevByteViewModel.kt:113]: this asynchronism in kotlin is still a bit confusing for me

Comment (joaobertacchi) [reply RepositoryPattern-Starter/app/src/main/java/com/example/android/devbyteviewer/viewmodels/DevByteViewModel.kt:113]: I think I have figured that out. The following declarations are equivalent:
```
    fun test1() = scope.launch { }
    fun test2(): Job {
        return scope.launch { }
    }
```

As `refreshDataFromNetwork()` returned a `Job` that was used for nothing, replacing that by `refreshDataFromRepository()` that returns `Unit` is ok.

**Pull request merged**